Gems/Atom/Tools/MaterialCanvas/Code/Source/Document/MaterialCanvasDocument.cpp

```diff
@@ -352,4 +372,337 @@ namespace MaterialCanvas
+        auto convertSlotTypeName = [](const AZStd::string& slotTypeName) -> AZStd::string
+        {
+            if (AZ::StringFunc::Equal(slotTypeName, "color"))
+            {
+                return "float4";
+            }
+
+            return slotTypeName;
+        };
+
+        // Disconnected input and property spots need to have their values converted into a format that can be injected into the shader.
+        auto convertSlotValue = [](const AZStd::any& slotValue) -> AZStd::string
+        {
+            if (auto v = AZStd::any_cast<const AZ::Color>(&slotValue))
+            {
+                return AZStd::string::format("float4(%g, %g, %g, %g)", v->GetR(), v->GetG(), v->GetB(), v->GetA());
+            }
+            if (auto v = AZStd::any_cast<const AZ::Vector4>(&slotValue))
+            {
+                return AZStd::string::format("float4(%g, %g, %g, %g)", v->GetX(), v->GetY(), v->GetZ(), v->GetW());
+            }
+            if (auto v = AZStd::any_cast<const AZ::Vector3>(&slotValue))
+            {
+                return AZStd::string::format("float3(%g, %g, %g)", v->GetX(), v->GetY(), v->GetZ());
+            }
+            if (auto v = AZStd::any_cast<const AZ::Vector2>(&slotValue))
+            {
+                return AZStd::string::format("float2(%g, %g)", v->GetX(), v->GetY());
+            }
+            if (auto v = AZStd::any_cast<const float>(&slotValue))
+            {
+                return AZStd::string::format("%g", *v);
+            }
+            if (auto v = AZStd::any_cast<const int>(&slotValue))
+            {
+                return AZStd::string::format("%i", *v);
+            }
+            if (auto v = AZStd::any_cast<const unsigned int>(&slotValue))
+            {
+                return AZStd::string::format("%u", *v);
+            }
+            return AZStd::string();
+        };
+
+        // Gather all of the common settings usually associated with a material canvas note to hand slot.
+        auto collectCommonSettings = [&](const AtomToolsFramework::DynamicNodeSettingsMap& settings)
+        {
+            collectSettingsAsSet(settings, "includePaths", includePaths);
+            collectSettingsAsSet(settings, "templatePaths", templatePaths);
+            collectSettingsAsVec(settings, "classDefinitions", classDefinitions);
+            collectSettingsAsVec(settings, "functionDefinitions", functionDefinitions);
+        };
+
+        // Each instruction block needs to have name substitutions performed on it. Variable names need to be unique per
+        // node and will be prepended with the node ID. Slot types and values will be determined based on connections and
+        // then substituted.
+        auto updateAndAddSlotInstructions =
+            [&](GraphModel::ConstNodePtr node, GraphModel::ConstSlotPtr slot, AZStd::vector<AZStd::string>& instructionsForSlot)
+        {
+            replaceStringsInVec("NODEID", AZStd::string::format("node%u", node->GetId()), instructionsForSlot);
```

Comment:
At some point we should probably let the user view and edit the ID for each node, so they can generate more debuggable shader code.

Reply:
Instead of using “node” I can use the sanitized version of the node name to generate the variable name prefixes as we make improvements